File: frontend/src/services/templates.ts
/**
 * Template browsing and self-service program start service.
 */

import { apiFetch } from './api';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface TemplateListItem {
  id: string;
  source: 'engine' | 'coach';
  name: string;
  description?: string;
  program_type?: string;
  difficulty_level?: string;
  duration_weeks: number;
  days_per_week: number;
  tags: string[];
  times_assigned: number;
}

export interface TemplateListResponse {
  templates: TemplateListItem[];
  total: number;
}

export interface ExercisePrescription {
  id: string;
  exercise_name: string;
  sets: number;
  reps?: number;
  reps_target?: number;
  weight_lbs?: number;
  rpe_target?: number;
  percentage_1rm?: number;
  rest_seconds?: number;
  notes?: string;
  coaching_cues?: string;
  exercise_order?: number;
}

export interface TodayWorkoutResponse {
  assignment_id: string;
  program_day_id: string;
  week_number: number;
  day_number: number;
  day_name: string;
  exercises: ExercisePrescription[];
  current_week: number;
  current_day: number;
  progress_percentage: number;
  progress_health: 'green' | 'yellow' | 'red';
}

export interface AssignmentState {
  id: string;
  program_id: string;
  client_id: string;
  assignment_name?: string;
  start_date: string;
  end_date?: string;
  status: string;
  current_week: number;
  current_day: number;
  workouts_completed: number;
  workouts_skipped: number;
  progress_percentage: number;
  progress_health: 'green' | 'yellow' | 'red';
  client_rating?: number;
  client_feedback?: string;
}

export interface SelfStartResponse {
  assignment_id: string;
  program_id: string;
  start_date: string;
  end_date: string;
  status: string;
}

export interface SetLogRequest {
  set_number: number;
  actual_reps?: number;
  actual_weight_lbs?: number;
  actual_rpe?: number;
  notes?: string;
}

export interface ExerciseLogRequest {
  program_day_exercise_id?: string;
  exercise_name: string;
  sets: SetLogRequest[];
}

export interface WorkoutLogExtendedRequest {
  assignment_id: string;
  program_day_id?: string;
  day_status: 'completed' | 'skipped' | 'partial';
  duration_minutes?: string;
  session_rating?: number;
  notes?: string;
  exercise_logs: ExerciseLogRequest[];
}

export interface WorkoutLogExtendedResponse {
  workout_log_id: string;
  assignment_id: string;
  program_day_id?: string;
  day_status?: string;
  duration_minutes?: string;
  session_rating?: number;
  exercise_logs: unknown[];
  current_week: number;
  current_day: number;
  assignment_status: string;
  progress_percentage: number;
  progress_health: string;
}

// ---------------------------------------------------------------------------
// Program detail (full tree for started programs)
// ---------------------------------------------------------------------------

export interface ProgramExercise {
  id: string;
  exercise_name: string;
  sets: number;
  reps: number;
  weight_lbs?: number;
  notes?: string;
}

export interface ProgramDay {
  id: string;
  day_number: number;
  name: string;
  exercises: ProgramExercise[];
}

export interface ProgramWeek {
  week_number: number;
  name: string;
  days: ProgramDay[];
}

export interface ProgramDetail {
  id: string;
  name: string;
  description?: string;
  duration_weeks: number;
  days_per_week: number;
  weeks: ProgramWeek[];
}

export async function getProgramDetail(programId: string): Promise<ProgramDetail> {
  return apiFetch<ProgramDetail>(`/programs/${programId}`);
}

// ---------------------------------------------------------------------------
// Template browsing
// ---------------------------------------------------------------------------

export interface TemplateFilters {
  source?: 'engine' | 'coach';
  difficulty?: string;
  duration_weeks?: number;
}

export async function listTemplates(filters: TemplateFilters = {}): Promise<TemplateListResponse> {
  const params = new URLSearchParams();
  if (filters.source) params.set('source', filters.source);
  if (filters.difficulty) params.set('difficulty', filters.difficulty);
  if (filters.duration_weeks) params.set('duration_weeks', String(filters.duration_weeks));
  const qs = params.toString();
  return apiFetch<TemplateListResponse>(`/programs/templates${qs ? `?${qs}` : ''}`);
}

export async function getTemplateDetail(source: 'engine' | 'coach', templateId: string): Promise<unknown> {
  return apiFetch<unknown>(`/programs/templates/${templateId}?source=${source}`);
}

export async function previewTemplate(
  templateId: string,
  body: { source: 'engine' | 'coach'; inputs?: Record<string, unknown>; movements?: unknown[] }
): Promise<unknown> {
  return apiFetch<unknown>(`/programs/templates/${templateId}/preview`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export async function selfStartProgram(
  templateId: string,
  body: {
    source: 'engine' | 'coach';
    inputs?: Record<string, unknown>;
    movements?: unknown[];
    start_date?: string;
    assignment_name?: string;
  }
): Promise<SelfStartResponse> {
  return apiFetch<SelfStartResponse>(`/programs/templates/${templateId}/start`, {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

// ---------------------------------------------------------------------------
// Assignment management
// ---------------------------------------------------------------------------

export async function getAssignment(assignmentId: string): Promise<AssignmentState> {
  return apiFetch<AssignmentState>(`/workouts/assignments/${assignmentId}`);
}

export async function getTodayWorkout(
  assignmentId: string,
  week?: number,
  day?: number,
): Promise<TodayWorkoutResponse> {
  const params = new URLSearchParams();
  if (week !== undefined) params.set('week', String(week));
  if (day !== undefined) params.set('day', String(day));
  const qs = params.toString();
  return apiFetch<TodayWorkoutResponse>(`/workouts/assignments/${assignmentId}/today${qs ? `?${qs}` : ''}`);
}

export interface DayLogSummary {
  workout_log_id: string;
  program_day_id: string | null;
  day_status: string;
  workout_date: string;
}

export async function getAssignmentLogs(assignmentId: string): Promise<DayLogSummary[]> {
  return apiFetch<DayLogSummary[]>(`/workouts/assignments/${assignmentId}/logs`);
}

export interface PreviousSetLog {
  set_number: number;
  actual_reps: number | null;
  actual_weight_lbs: number | null;
  actual_rpe: number | null;
  notes: string | null;
}

export interface PreviousExerciseLog {
  exercise_name: string;
  program_day_exercise_id: string | null;
  sets: PreviousSetLog[];
}

export interface WorkoutFullDetail {
  workout_log_id: string;
  day_status: string | null;
  duration_minutes: string | null;
  session_rating: number | null;
  notes: string | null;
  exercise_logs: PreviousExerciseLog[];
}

export async function getWorkoutDetail(workoutLogId: string): Promise<WorkoutFullDetail> {
  return apiFetch<WorkoutFullDetail>(`/workouts/${workoutLogId}/detail`);
}

export async function logWorkout(body: WorkoutLogExtendedRequest): Promise<WorkoutLogExtendedResponse> {
  return apiFetch<WorkoutLogExtendedResponse>('/workouts/log', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export async function submitFeedback(
  assignmentId: string,
  body: { client_rating?: number; client_feedback?: string }
): Promise<AssignmentState> {
  return apiFetch<AssignmentState>(`/workouts/assignments/${assignmentId}/feedback`, {
    method: 'PATCH',
    body: JSON.stringify(body),
  });
}

export async function overrideProgress(
  assignmentId: string,
  body: { current_week?: number; current_day?: number; status?: string }
): Promise<AssignmentState> {
  return apiFetch<AssignmentState>(`/workouts/assignments/${assignmentId}/progress`, {
    method: 'PATCH',
    body: JSON.stringify(body),
  });
}

File: backend/app/api/templates.py
"""
Program template browsing and client self-service start flow.

GET  /programs/templates                    — list engine + coach templates
GET  /programs/templates/{id}               — template detail
POST /programs/templates/{id}/preview       — dry-run generation
POST /programs/templates/{id}/start         — generate copy + assign to self
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import and_, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.program import Program, ProgramDay, ProgramWeek
from app.models.user import User
from app.schemas.program import (
    DayDetail,
    ExerciseDetail,
    ProgramDetailResponse,
    WeekDetail,
)
from app.schemas.program_self_start import (
    SelfStartRequest,
    SelfStartResponse,
    TemplateListItem,
    TemplateListResponse,
)
from app.services import engine_client

router = APIRouter()


# ---------------------------------------------------------------------------
# GET /programs/templates
# ---------------------------------------------------------------------------


@router.get("/", response_model=TemplateListResponse, summary="List available program templates")
async def list_templates(
    source: str | None = None,  # "engine" | "coach" | None (all)
    difficulty: str | None = None,
    duration_weeks: int | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Returns both engine-defined programs and coach-created templates visible
    to the current user's subscription.
    """
    items: list[TemplateListItem] = []

    # --- Engine templates ---
    if source in (None, "engine"):
        try:
            engine_defs = await engine_client.list_program_definitions()
        except HTTPException:
            engine_defs = []

        for defn in engine_defs:
            dw = defn.get("duration_weeks") or defn.get("weeks")
            dpw = defn.get("days_per_week") or defn.get("days")
            if duration_weeks and dw != duration_weeks:
                continue
            if difficulty and defn.get("difficulty_level") != difficulty:
                continue
            items.append(
                TemplateListItem(
                    id=str(defn.get("id") or defn.get("program_id", "")),
                    source="engine",
                    name=defn.get("name", ""),
                    description=defn.get("description"),
                    program_type=defn.get("program_type"),
                    difficulty_level=defn.get("difficulty_level"),
                    duration_weeks=dw or 0,
                    days_per_week=dpw or 0,
                    tags=defn.get("tags", []),
                    times_assigned=defn.get("times_assigned", 0),
                )
            )

    # --- Coach templates (subscription-scoped + public) ---
    if source in (None, "coach"):
        query = (
            select(Program)
            .where(
                and_(
                    Program.is_template == True,
                    Program.archived_at.is_(None),
                    or_(
                        Program.subscription_id == current_user.subscription_id,
                        Program.is_public == True,
                    ),
                )
            )
            .order_by(Program.created_at.desc())
        )
        if difficulty:
            query = query.where(Program.difficulty_level == difficulty)
        if duration_weeks:
            query = query.where(Program.duration_weeks == duration_weeks)

        result = await db.execute(query)
        programs = result.scalars().all()

        for p in programs:
            items.append(
                TemplateListItem(
                    id=str(p.id),
                    source="coach",
                    name=p.name,
                    description=p.description,
                    program_type=p.program_type,
                    difficulty_level=p.difficulty_level,
                    duration_weeks=p.duration_weeks,
                    days_per_week=p.days_per_week,
                    tags=p.tags or [],
                    times_assigned=p.times_assigned or 0,
                )
            )

    return TemplateListResponse(templates=items, total=len(items))


# ---------------------------------------------------------------------------
# GET /programs/templates/{id}
# ---------------------------------------------------------------------------


@router.get("/{template_id}", summary="Template detail")
async def get_template(
    template_id: str,
    source: str = "coach",  # "engine" | "coach"
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    For engine templates: proxies to TrainGen Engine definition detail.
    For coach templates: returns full program tree from DB.
    """
    if source == "engine":
        return await engine_client.get_program_definition(template_id)

    from uuid import UUID

    try:
        template_uuid = UUID(template_id)
    except ValueError:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid template ID")

    result = await db.execute(
        select(Program)
        .options(
            selectinload(Program.weeks)
            .selectinload(ProgramWeek.days)
            .selectinload(ProgramDay.exercises)
        )
        .where(
            and_(
                Program.id == template_uuid,
                Program.is_template == True,
                Program.archived_at.is_(None),
                or_(
                    Program.subscription_id == current_user.subscription_id,
                    Program.is_public == True,
                ),
            )
        )
    )
    program = result.scalar_one_or_none()
    if not program:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Template not found")

    weeks = []
    for week in sorted(program.weeks, key=lambda w: w.week_number):
        days = []
        for day in sorted(week.days, key=lambda d: d.day_number):
            exercises = [
                ExerciseDetail(
                    id=str(ex.id),
                    exercise_name=ex.exercise_name or "",
                    sets=ex.sets,
                    reps=ex.reps or ex.reps_target or 0,
                    weight_lbs=ex.weight_lbs or ex.load_value,
                    percentage_1rm=int(ex.percentage_1rm) if ex.percentage_1rm else None,
                    notes=ex.notes or "",
                )
                for ex in sorted(day.exercises, key=lambda e: (e.exercise_order or 0))
            ]
            days.append(
                DayDetail(
                    day_number=day.day_number,
                    name=day.name,
                    suggested_day_of_week=(
                        str(day.suggested_day_of_week) if day.suggested_day_of_week else None
                    ),
                    exercises=exercises,
                )
            )
        weeks.append(
            WeekDetail(week_number=week.week_number, name=week.name or f"Week {week.week_number}", days=days)
        )

    return ProgramDetailResponse(
        id=str(program.id),
        subscription_id=str(program.subscription_id) if program.subscription_id else None,
        created_by_user_id=str(program.created_by_user_id) if program.created_by_user_id else None,
        name=program.name,
        description=program.description,
        builder_type=program.builder_type,
        algorithm_version=program.algorithm_version,
        duration_weeks=program.duration_weeks,
        days_per_week=program.days_per_week,
        is_template=program.is_template,
        is_public=program.is_public,
        times_assigned=program.times_assigned or 0,
        status=program.status,
        created_at=program.created_at.isoformat(),
        updated_at=program.updated_at.isoformat(),
        input_data=program.input_data or {},
        calculated_data=program.calculated_data or {},
        weeks=weeks,
    )


# ---------------------------------------------------------------------------
# POST /programs/templates/{id}/preview
# ---------------------------------------------------------------------------


@router.post("/{template_id}/preview", summary="Dry-run program generation")
async def preview_template(
    template_id: str,
    body: SelfStartRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Dry-run: call the engine or internal generator and return the full program
    structure without persisting anything.
    """
    if body.source == "engine":
        payload = dict(body.inputs)
        if body.movements:
            payload.setdefault("movements", body.movements)
        payload.setdefault("program_id", template_id)
        return await engine_client.generate_plan(payload)

    # Coach template preview: load template and return its current structure
    return await get_template(template_id=template_id, source="coach", db=db, current_user=current_user)


# ---------------------------------------------------------------------------
# POST /programs/templates/{id}/start
# ---------------------------------------------------------------------------


@router.post(
    "/{template_id}/start",
    response_model=SelfStartResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Generate program copy and assign to self",
)
async def start_template(
    template_id: str,
    body: SelfStartRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Generates a full copy of the program (engine or coach template),
    creates a ClientProgramAssignment for the current user, and schedules days.
    """
    from datetime import date as date_type

    from app.services.assignment_service import self_start_program

    start_date = body.start_date or date_type.today()

    inputs = dict(body.inputs)
    if body.movements:
        inputs.setdefault("movements", body.movements)

    assignment = await self_start_program(
        source=body.source,
        template_id=template_id,
        inputs=inputs,
        start_date=start_date,
        current_user=current_user,
        db=db,
        assignment_name=body.assignment_name,
    )

    await db.commit()
    await db.refresh(assignment)

    return SelfStartResponse(
        assignment_id=str(assignment.id),
        program_id=str(assignment.program_id),
        start_date=assignment.start_date,
        end_date=assignment.end_date,
        status=assignment.status,
    )
